aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs

```diff
@@ -0,0 +1,648 @@
+
+impl Aggregator {
+    /// Implements logic for adding to an aggregator.
+    fn add(&mut self, value: u128) -> PartialVMResult<()> {
+        // At this point, aggregator holds a positive delta or knows the value.
+        // Hence, we can add, of course checking overflow condition.
+        match addition(self.value, value, self.limit) {
+            Some(result) => {
+                self.value = result;
+                Ok(())
+            }
+            None => Err(abort_error(
+                format!("aggregator's value overflowed when adding {}", value),
+                EAGGREGATOR_OVERFLOW,
+            )),
+        }
+    }
+
+    /// Implements logic for subtracting from an aggregator.
+    fn sub(&mut self, value: u128) -> PartialVMResult<()> {
+        // First, we check if the value can be subtracted in theory, that is
+        // if `value <= limit`. If not, we can immediately abort.
+        if value > self.limit {
+            return Err(abort_error(
+                format!("aggregator's value underflowed when subtracting more than aggregator can hold ({} > limit)", value),
+                EAGGREGATOR_UNDERFLOW,
+            ));
+        }
+
+        // We can subtract. Now consider each possible state of aggregator
+        // separately.
+        match self.state {
+            AggregatorState::Data => {
+                // Aggregator knows the value, therefore we can subtract
+                // checking we don't drop below zero.
+                match subtraction(self.value, value) {
+                    Some(result) => {
+                        self.value = result;
+                        Ok(())
+                    }
+                    None => Err(abort_error(
+                        format!(
+                            "aggregator's value underflowed when subtracting {} from {}",
+                            value, self.value
+                        ),
+                        EAGGREGATOR_UNDERFLOW,
+                    )),
+                }
+            }
+            // Since `sub` is a barrier, we never encounter delta during
+            // subtraction. This will change in future.
+            // TODO: implement this properly once `sub` stops being a barrier.
+            AggregatorState::PositiveDelta => {
+                unreachable!("subtraction always materializes the value")
+            }
+        }
+    }
+
+    /// Materializes the value of the aggregator, changing its state.
+    fn materialize(
+        &mut self,
+        context: &NativeAggregatorContext,
+        handle: &TableHandle,
+    ) -> PartialVMResult<()> {
+        // If aggregator has already been materialized, return immediately.
+        if self.state == AggregatorState::Data {
+            return Ok(());
+        }
+
+        // Otherwise, we have a delta and have to go to storage.
+        let key_bytes = serialize(&self.key);
+        match context.resolver.resolve_table_entry(handle, &key_bytes) {
```

Comment:
let bytes = ...map_err(extension_error(...))?.map_err(extension_error(...))?

Should work, but I like these chains, many people don't. Whatever you and others prefer. I just don't like nested matches everything becomes to indented OCD